# Proyecto: Exploración y visualización de datos.

### 1. Importación de librerías y archivos

In [2]:
# Importar las librerías necesarias según el análisis que se vaya a realizar
# Librería para manejo de datos convensional
import pandas as pd
#librería para creación de gráficos
import matplotlib.pyplot as plt
# Librería para perfilamiento
#import pandas_profiling

In [6]:
# cargar los datos en csv
data= pd.read_csv('empleadosTecnologiaAlpes.csv', sep=';',encoding='latin-1')
# Visualizar los datos
data

Edad Retirado       ViajesDeNegocio  TarifaSalarial  \
0       41       Si     Viaja-Pocas-Veces            1102   
1       49       No  Viaja-Frecuentemente             279   
2       37       Si     Viaja-Pocas-Veces            1373   
3       33       No  Viaja-Frecuentemente            1392   
4       27       No     Viaja-Pocas-Veces             591   
...    ...      ...                   ...             ...   
1487    26       Si     Viaja-Pocas-Veces            1330   
1488    23       Si     Viaja-Pocas-Veces            1243   
1489    23       Si     Viaja-Pocas-Veces            1320   
1490    26       Si     Viaja-Pocas-Veces             920   
1491    23       Si     Viaja-Pocas-Veces            1320   

                    Departamento  DistanciaDesdeCasa  Educacion  \
0                         Ventas                   1          2   
1     Investigacion & Desarrollo                   8          1   
2     Investigacion & Desarrollo                   2          2   
3     Investigacion & Desarrollo                   3          4   
4     Investigacion & Desarrollo                   2          1   
...                          ...                 ...        ...   
1487  Investigacion & Desarrollo                  21          3   
1488  Investigacion & Desarrollo                   6          3   
1489  Investigacion & Desarrollo                   8          1   
1490              Talento Humano                  20          2   
1491  Investigacion & Desarrollo                   8          1   

           CampoEducacion  EmpleadosACargo  NumeroEmpleado  ...  \
0     Ciencias de la vida              1.0               1  ...   
1     Ciencias de la vida              1.0               2  ...   
2                    Otro              1.0               4  ...   
3     Ciencias de la vida              1.0               5  ...   
4                Medicina              1.0               7  ...   
...                   ...              ...             ...  ...   
1487             Medicina              1.0            1107  ...   
1488  Ciencias de la vida              1.0             811  ...   
1489             Medicina              1.0            1684  ...   
1490             Medicina              1.0            1818  ...   
1491             Medicina              1.0            1684  ...   

      HorasEstandar PlanSalarioComplementario  TotalAnniosTrabajados  \
0                80                         0                      8   
1                80                         1                     10   
2                80                         0                      7   
3                80                         0                      8   
4                80                         1                      6   
...             ...                       ...                    ...   
1487             80                         1                      1   
1488             80                         2                      1   
1489             80                         0                      5   
1490             80                         0                      6   
1491             80                         0                      5   

      FormacionUltimoAnnio  BalanceTrabajoVida FechaIngresoEmpresa  \
0                        0                   1          26/07/2015   
1                        3                   3          15/02/2011   
2                        3                   3           2/06/2021   
3                        3                   3          26/09/2013   
4                        3                   3           2/03/2019   
...                    ...                 ...                 ...   
1487                     0                   2          14/09/2020   
1488                     2                   3           5/09/2021   
1489                     2                   3          28/01/2016   
1490                     3                   3          16/11/2016   
1491                     2  

### 2. Análisis de la fuente de los datos

### 2.1. Registros y atributos
El número de registros y atributos nos dará una guía del correcto cargue de los datos, podemos utilizar la función shape para la revisión de este dato

In [4]:
# Tamaño del dataset (filas, columnas)
data.shape

(1492, 36)

In [5]:
# Revisión de los atributos
data.columns

Index(['Edad', 'Retirado', 'ViajesDeNegocio', 'TarifaSalarial', 'Departamento',
       'DistanciaDesdeCasa', 'Educacion', 'CampoEducacion', 'EmpleadosACargo',
       'NumeroEmpleado', 'NivelSatisfaccion', 'Genero', 'SalarioPorHora',
       'NivelInvolucramiento', 'NivelResponsabilidad', 'Rol',
       'SatisfaccionTrabajo', 'EstadoCivil', 'SalarioMensual', 'TarifaMensual',
       'NumEmpresasDondeTrabaja', 'MasDe18Horas', 'HorasExtra',
       'PorcentajeIncrementoSalarioUltimoAnnio', 'NivelRendimiento',
       'SatisfaccionEmpresa', 'HorasEstandar', 'PlanSalarioComplementario',
       'TotalAnniosTrabajados', 'FormacionUltimoAnnio', 'BalanceTrabajoVida',
       'FechaIngresoEmpresa', 'AnniosEnEmpresa', 'AnniosRolActual',
       'AnniosDesdeUltimaPromocion', 'AnniosActualDirector'],
      dtype='object')

### 2.2. Determinar el identificador único de la fuente de datos y validar sus propiedades


In [7]:
# Revisar los registros que contienen vacíos en el atributo orden_id
data['NumeroEmpleado'].isna().sum()

0

In [8]:
# revisar que no existan valores duplicados en el atributo order_id
len(data['NumeroEmpleado'].unique())

1470

En este punto recordemos que la fuente tiene 1492 registros, de esta manera identificamos que hay 22 registros duplicados y es lo que queremos validar a continuación.

In [10]:
# contar los registros duplicados del atributo orden_id
data.duplicated('NumeroEmpleado').sum()

22

En la validación del identificador vemos que no tenemos problemas de registros vacíos o nulos, pero si de duplicados. Pasamos a revisar el tema de los duplicados.

In [11]:
# visualizar los datos duplicados de acuerdo con el identificador - orden_id
data[data.duplicated('NumeroEmpleado')].head()

Edad Retirado       ViajesDeNegocio  TarifaSalarial  \
1201    23       Si     Viaja-Pocas-Veces             427   
1438    23       Si  Viaja-Frecuentemente             638   
1470    26       Si     Viaja-Pocas-Veces            1357   
1471    20       Si  Viaja-Frecuentemente             871   
1472    26       Si     Viaja-Pocas-Veces            1449   

                    Departamento  DistanciaDesdeCasa  Educacion  \
1201                      Ventas                   7          3   
1438                      Ventas                   9          3   
1470  Investigacion & Desarrollo                  25          3   
1471  Investigacion & Desarrollo                   6          3   
1472  Investigacion & Desarrollo                  16          4   

           CampoEducacion  EmpleadosACargo  NumeroEmpleado  ...  \
1201  Ciencias de la vida              1.0            1702  ...   
1438             Medicina              1.0            2023  ...   
1470  Ciencias de la vida              1.0              55  ...   
1471  Ciencias de la vida              1.0             137  ...   
1472             Medicina              1.0             394  ...   

      HorasEstandar PlanSalarioComplementario  TotalAnniosTrabajados  \
1201             80                         1                      3   
1438             80                         1                      1   
1470             80                         0                      1   
1471             80                         0                      1   
1472             80                         1                      5   

      FormacionUltimoAnnio  BalanceTrabajoVida FechaIngresoEmpresa  \
1201                     2                   3           5/06/2018   
1438                     3                   2          25/04/2020   
1470                     2                   2           3/04/2020   
1471                     5                   3          28/05/2020   
1472                     2                   3          18/10/2018   

      AnniosEnEmpresa AnniosRolActual  AnniosDesdeUltimaPromocion  \
1201                3               2                           0   
1438                1               0                           1   
1470                1               0                           0   
1471                1               0                           1   
1472                3               2                           0   

      AnniosActualDirector  
1201                     2  
1438                     0  
1470                     1  
1471                     0  
1472                     2  

[5 rows x 36 columns]

Con este resultado podemos ver algunas de las órdenes duplicadas, pero es necesario identificar cuáles son esos datos duplicados y sus características. 

In [12]:
# encontrar los registros en los que el identificador de la orden está duplicado pero los datos de la orden no.
data[data.duplicated('NumeroEmpleado') & ~data.duplicated()]

Edad Retirado       ViajesDeNegocio  TarifaSalarial Departamento  \
1438    23       Si  Viaja-Frecuentemente             638       Ventas   

      DistanciaDesdeCasa  Educacion CampoEducacion  EmpleadosACargo  \
1438                   9          3       Medicina              1.0   

      NumeroEmpleado  ...  HorasEstandar PlanSalarioComplementario  \
1438            2023  ...             80                         1   

      TotalAnniosTrabajados  FormacionUltimoAnnio  BalanceTrabajoVida  \
1438                      1                     3                   2   

     FechaIngresoEmpresa  AnniosEnEmpresa AnniosRolActual  \
1438          25/04/2020                1               0   

      AnniosDesdeUltimaPromocion  AnniosActualDirector  
1438                           1                     0  

[1 rows x 36 columns]

Si bien es cierto que se identificaron estos registros como duplicados en el atributo orden_id, la siguiente tarea es analizar el contenido de las otras variables para terminar de comprender el problema.

In [17]:
# revisar uno de los datos donde la orden está duplicada
data[data['NumeroEmpleado']==2023]

Edad Retirado       ViajesDeNegocio  TarifaSalarial Departamento  \
1213    23       Si  Viaja-Frecuentemente             638       Ventas   
1438    23       Si  Viaja-Frecuentemente             638       Ventas   

      DistanciaDesdeCasa  Educacion CampoEducacion  EmpleadosACargo  \
1213                   9          3       Mercadeo              1.0   
1438                   9          3       Medicina              1.0   

      NumeroEmpleado  ...  HorasEstandar PlanSalarioComplementario  \
1213            2023  ...             80                         1   
1438            2023  ...             80                         1   

      TotalAnniosTrabajados  FormacionUltimoAnnio  BalanceTrabajoVida  \
1213                      1                     3                   2   
1438                      1                     3                   2   

     FechaIngresoEmpresa  AnniosEnEmpresa AnniosRolActual  \
1213          25/04/2020                1               0   
1438          25/04/2020                1               0   

      AnniosDesdeUltimaPromocion  AnniosActualDirector  
1213                           1                     0  
1438                           1                     0  

[2 rows x 36 columns]

In [18]:
# revisar uno de los datos donde la orden está duplicada
data[data['NumeroEmpleado']==1702]

Edad Retirado    ViajesDeNegocio  TarifaSalarial Departamento  \
585     23       Si  Viaja-Pocas-Veces             427       Ventas   
1201    23       Si  Viaja-Pocas-Veces             427       Ventas   

      DistanciaDesdeCasa  Educacion       CampoEducacion  EmpleadosACargo  \
585                    7          3  Ciencias de la vida              1.0   
1201                   7          3  Ciencias de la vida              1.0   

      NumeroEmpleado  ...  HorasEstandar PlanSalarioComplementario  \
585             1702  ...             80                         1   
1201            1702  ...             80                         1   

      TotalAnniosTrabajados  FormacionUltimoAnnio  BalanceTrabajoVida  \
585                       3                     2                   3   
1201                      3                     2                   3   

     FechaIngresoEmpresa  AnniosEnEmpresa AnniosRolActual  \
585            5/06/2018                3               2   
1201           5/06/2018                3               2   

      AnniosDesdeUltimaPromocion  AnniosActualDirector  
585                            0                     2  
1201                           0                     2  

[2 rows x 36 columns]

Solo hay dos registro que comparten la misma informacion.

### 2.3. Duplicados de la fuente de datos

Una vez encontrado el identificador único podemos revisar cuáles son los registros que se están duplicando para realizar las correcciones necesarias. La existencia de estos registros puede deberse a diferentes razones, puede ser simplemente un error en la captura de información o algo más complejo, como que existan ordenes duplicadas en el sistema fuente. Por esto, la importancia de determinar los registros al parecer duplicados.

In [19]:
# contar los registros duplicados de toda la fuente
data.duplicated().sum()

21

En esta validación vemos que hay menos registros duplicados que cuando contamos valores duplicados en el identificador. Veámos las características de los registros con identificador duplicado.

In [20]:
# revisar el número de duplicados de todos los atributos con la función count
data[data.duplicated()].count()

Edad                                      21
Retirado                                  21
ViajesDeNegocio                           21
TarifaSalarial                            21
Departamento                              21
DistanciaDesdeCasa                        21
Educacion                                 21
CampoEducacion                            21
EmpleadosACargo                           21
NumeroEmpleado                            21
NivelSatisfaccion                         21
Genero                                    21
SalarioPorHora                            21
NivelInvolucramiento                      21
NivelResponsabilidad                      21
Rol                                       21
SatisfaccionTrabajo                       21
EstadoCivil                               21
SalarioMensual                            21
TarifaMensual                             21
NumEmpresasDondeTrabaja                   21
MasDe18Horas                              21
HorasExtra

En este punto es importante reflexionar sobre el resultado obtenido y determinar si es necesario hacer otros análisis para comprender mejor el problema y poderlo reportar.

In [21]:
# revisar el contenido de los registros duplicados
data[data.duplicated()].head()

Edad Retirado       ViajesDeNegocio  TarifaSalarial  \
1201    23       Si     Viaja-Pocas-Veces             427   
1470    26       Si     Viaja-Pocas-Veces            1357   
1471    20       Si  Viaja-Frecuentemente             871   
1472    26       Si     Viaja-Pocas-Veces            1449   
1473    26       Si     Viaja-Pocas-Veces             950   

                    Departamento  DistanciaDesdeCasa  Educacion  \
1201                      Ventas                   7          3   
1470  Investigacion & Desarrollo                  25          3   
1471  Investigacion & Desarrollo                   6          3   
1472  Investigacion & Desarrollo                  16          4   
1473                      Ventas                   4          4   

           CampoEducacion  EmpleadosACargo  NumeroEmpleado  ...  \
1201  Ciencias de la vida              1.0            1702  ...   
1470  Ciencias de la vida              1.0              55  ...   
1471  Ciencias de la vida              1.0             137  ...   
1472             Medicina              1.0             394  ...   
1473             Mercadeo              1.0             401  ...   

      HorasEstandar PlanSalarioComplementario  TotalAnniosTrabajados  \
1201             80                         1                      3   
1470             80                         0                      1   
1471             80                         0                      1   
1472             80                         1                      5   
1473             80                         0                      8   

      FormacionUltimoAnnio  BalanceTrabajoVida FechaIngresoEmpresa  \
1201                     2                   3           5/06/2018   
1470                     2                   2           3/04/2020   
1471                     5                   3          28/05/2020   
1472                     2                   3          18/10/2018   
1473                     0                   3          26/09/2013   

      AnniosEnEmpresa AnniosRolActual  AnniosDesdeUltimaPromocion  \
1201                3               2                           0   
1470                1               0                           0   
1471                1               0                           1   
1472                3               2                           0   
1473                8               7                           7   

      AnniosActualDirector  
1201                     2  
1470                     1  
1471                     0  
1472                     2  
1473                     4  

[5 rows x 36 columns]

In [22]:
# revisar el contenido de uno de los registros duplicados
data[data['NumeroEmpleado']==137]

Edad Retirado       ViajesDeNegocio  TarifaSalarial  \
102     20       Si  Viaja-Frecuentemente             871   
1471    20       Si  Viaja-Frecuentemente             871   

                    Departamento  DistanciaDesdeCasa  Educacion  \
102   Investigacion & Desarrollo                   6          3   
1471  Investigacion & Desarrollo                   6          3   

           CampoEducacion  EmpleadosACargo  NumeroEmpleado  ...  \
102   Ciencias de la vida              1.0             137  ...   
1471  Ciencias de la vida              1.0             137  ...   

      HorasEstandar PlanSalarioComplementario  TotalAnniosTrabajados  \
102              80                         0                      1   
1471             80                         0                      1   

      FormacionUltimoAnnio  BalanceTrabajoVida FechaIngresoEmpresa  \
102                      5                   3          28/05/2020   
1471                     5                   3          28/05/2020   

      AnniosEnEmpresa AnniosRolActual  AnniosDesdeUltimaPromocion  \
102                 1               0                           1   
1471                1               0                           1   

      AnniosActualDirector  
102                      0  
1471                     0  

[2 rows x 36 columns]

Con este hallazgo ya es más fácil reportar los casos encontrados.

### 3. Análisis de los atributos


### 3.1. Tipos de datos

Debemos identificar en la fuente el tipo de datos de las variables que estamos analizando, para poder establecer su formato y su dominio.

In [23]:
#identificar los tipos de datos
data.dtypes

Edad                                        int64
Retirado                                   object
ViajesDeNegocio                            object
TarifaSalarial                              int64
Departamento                               object
DistanciaDesdeCasa                          int64
Educacion                                   int64
CampoEducacion                             object
EmpleadosACargo                           float64
NumeroEmpleado                              int64
NivelSatisfaccion                           int64
Genero                                     object
SalarioPorHora                              int64
NivelInvolucramiento                        int64
NivelResponsabilidad                        int64
Rol                                        object
SatisfaccionTrabajo                         int64
EstadoCivil                                object
SalarioMensual                              int64
TarifaMensual                               int64


### 3.2. Dominio

Debemos identificar en el diccionario de datos el dominio y corroborarlo con los datos.

In [24]:
#identificar el dominio del atributo precio, primero el valor mínimo
data['Edad'].min()

18

In [25]:
#identificar el dominio del atributo precio, en este punto el valor máximo
data['Edad'].max()

60

Vemos como el dominio de Edad está entre los valores 18 y 60.

### 3.3. Patrones

Debemos identificar los patrones de algunos de los atributos. En este caso el formato de la columna fecha_envio_límite y el código postal, por lo tanto vamos a realizar las siguientes actividades:

* Revisar formato del atributo fecha.
* Entender el formato de código postal.

In [26]:
#visualizar los datos que se van a corregir
data[['FechaIngresoEmpresa']].head()

FechaIngresoEmpresa
0          26/07/2015
1          15/02/2011
2           2/06/2021
3          26/09/2013
4           2/03/2019

**Fecha:** la fecha posee el siguiente formato:
* dd/mm/AAAA

Donde dd: es día con 1 o 2 dígitos, mm: los dos digitos del mes, AAAA: año con 4 dígitos.

### 3.4. Atributos vacíos

Uno de los pasos más importantes para el negocio es identificar qué atributos están vacíos, ya que esto puede alterar los resultados de los análisis que se están presentando, e incluso, en algunos casos, generar alguna pérdida económica si no está bien identificado.

In [27]:
# Revisar los atributos que contienen vacíos
data.isna().sum()

Edad                                       0
Retirado                                   0
ViajesDeNegocio                            0
TarifaSalarial                             0
Departamento                               0
DistanciaDesdeCasa                         0
Educacion                                  0
CampoEducacion                             0
EmpleadosACargo                           20
NumeroEmpleado                             0
NivelSatisfaccion                          0
Genero                                     7
SalarioPorHora                             0
NivelInvolucramiento                       0
NivelResponsabilidad                       0
Rol                                        0
SatisfaccionTrabajo                        0
EstadoCivil                               10
SalarioMensual                             0
TarifaMensual                              0
NumEmpresasDondeTrabaja                    0
MasDe18Horas                               0
HorasExtra

In [28]:
#porcentaje de nulos por variable
data.isna().sum()/len(data)

Edad                                      0.000000
Retirado                                  0.000000
ViajesDeNegocio                           0.000000
TarifaSalarial                            0.000000
Departamento                              0.000000
DistanciaDesdeCasa                        0.000000
Educacion                                 0.000000
CampoEducacion                            0.000000
EmpleadosACargo                           0.013405
NumeroEmpleado                            0.000000
NivelSatisfaccion                         0.000000
Genero                                    0.004692
SalarioPorHora                            0.000000
NivelInvolucramiento                      0.000000
NivelResponsabilidad                      0.000000
Rol                                       0.000000
SatisfaccionTrabajo                       0.000000
EstadoCivil                               0.006702
SalarioMensual                            0.000000
TarifaMensual                  

Con el código anterior es posible revisar cuáles atributos contienen nulos y cuáles no, además podemos ver el porcentaje de valores nulos por atributo. Hay que tener en cuenta que tenemos vacíos en variables fundamentales como el valor del flete y el nombre de categoría de producto, la pregunta es si esto es o no un error, el único que lo sabe es el negocio, así que debemos reportarlo y validarlo.

### 3.5. Uso de estadística

Apoyado en el uso de estadística descriptiva, (i.e., medidas de tendencia central y dispersión), es posible entender las distribuciones y patrones de los atributos. De igual manera, identificar valores faltantes, valores atípicos entre otros, utilizando estadísticos como el mínimo, el máximo, y la media.

In [29]:
# Estadísticas básicas para las variables numéricas
data.describe()

Edad  TarifaSalarial  DistanciaDesdeCasa    Educacion  \
count  1492.000000     1492.000000         1492.000000  1492.000000   
mean     36.730563      803.034853            9.221850     2.910188   
std       9.209743      403.481585            8.107823     1.022210   
min      18.000000      102.000000            1.000000     1.000000   
25%      30.000000      465.000000            2.000000     2.000000   
50%      35.000000      803.000000            7.000000     3.000000   
75%      43.000000     1157.250000           14.000000     4.000000   
max      60.000000     1499.000000           29.000000     5.000000   

       EmpleadosACargo  NumeroEmpleado  NivelSatisfaccion  SalarioPorHora  \
count           1472.0     1492.000000        1492.000000     1492.000000   
mean               1.0     1023.347185           2.723861       65.842493   
std                0.0      600.863591           1.094536       20.334559   
min                1.0        1.000000           1.000000       30.000000   
25%                1.0      492.750000           2.000000       48.000000   
50%                1.0     1016.500000           3.000000       66.000000   
75%                1.0     1554.250000           4.000000       83.250000   
max                1.0     2068.000000           4.000000      100.000000   

       NivelInvolucramiento  NivelResponsabilidad  ...  SatisfaccionEmpresa  \
count           1492.000000           1492.000000  ...          1492.000000   
mean               2.725871              2.050268  ...             2.708445   
std                0.713425              1.105207  ...             1.080333   
min                1.000000              1.000000  ...             1.000000   
25%                2.000000              1.000000  ...             2.000000   
50%                3.000000              2.000000  ...             3.000000   
75%                3.000000              3.000000  ...             4.000000   
max                4.000000              5.000000  ...             4.000000   

       HorasEstandar  PlanSalarioComplementario  TotalAnniosTrabajados  \
count         1492.0                1492.000000            1492.000000   
mean            80.0                   0.788204              11.168231   
std              0.0                   0.849931               7.783918   
min             80.0                   0.000000               0.000000   
25%             80.0                   0.000000               6.000000   
50%             80.0                   1.000000              10.000000   
75%             80.0                   1.000000              15.000000   
max             80.0                   3.000000              40.000000   

       FormacionUltimoAnnio  BalanceTrabajoVida  AnniosEnEmpresa  \
count           1492.000000         1492.000000      1492.000000   
mean               2.794906            2.759383         6.867962   
std                1.291057            0.704788         5.911656   
min                0.000000            1.000000         0.000000   
25%                2.000000            2.000000         3.000000   
50%                3.000000            3.000000         5.000000   
75%                3.000000            3.000000         9.000000   
max                6.000000            4.000000        40.000000   

       AnniosRolActual  AnniosDesdeUltimaPromocion  AnniosActualDirector  
count      1492.000000                 1492.000000           1492.000000  
mean          4.197721                    2.167560              4.087131  
std           3.616479                    3.207774              3.560408  
min           0.000000                    0.000000              0.000000  
25%           2.000000                    0.000000              2.000000  
50%           3.000000                    1.000000              3.000000  
75%           7.000000                    3.000000              7.000000  
max          18.000000                   15.000000             17.000000  

[8 rows x 26

In [31]:
# consultar el número de registros con precios negativos
len(data[data['Edad']<=0])

0

### 3.6. Estadísticas sobre datos categóricos

Con los atributos que son de tipo string y tienen categorías, como por ejemplo el departamento de estudio, se puede ver rápidamente si hay problemas.

In [32]:
#identificar las columnas de tipo string, si no se toma de la descripción inicial de la fuente de datos
data.select_dtypes(include='object').columns

Index(['Retirado', 'ViajesDeNegocio', 'Departamento', 'CampoEducacion',
       'Genero', 'Rol', 'EstadoCivil', 'MasDe18Horas', 'HorasExtra',
       'FechaIngresoEmpresa'],
      dtype='object')

Al revisar este resultado, ratificamos el hecho de que hay problemas de tipos de datos, los cuales debemos corregir. Este es el caso de orden_id, producto_id, vendedor_id, fecha_envío_limite. Vamos a continuar con los que en realidad son categóricos.

In [34]:
# contar los valores de cada atributo categórico
for i in ['Retirado', 'ViajesDeNegocio','Departamento','Rol']:
    print('Atributo: '+ i)
    print(data[i].value_counts())

Atributo: Retirado
No    1233
Si     259
Name: Retirado, dtype: int64
Atributo: ViajesDeNegocio
Viaja-Pocas-Veces       1057
Viaja-Frecuentemente     284
No-Viaja                 151
Name: ViajesDeNegocio, dtype: int64
Atributo: Departamento
Investigacion & Desarrollo    974
Ventas                        451
Talento Humano                 65
Venta                           1
Venta                           1
Name: Departamento, dtype: int64
Atributo: Rol
Ejecutivo de ventas              329
Cientifico de investigacion      297
Tecnico de laboratorio           267
Director de manufactura          145
Representante atencion medica    131
Director                         102
Representante de ventas           87
Director de investigacion         80
Talento Humano                    54
Name: Rol, dtype: int64


Se puede ver que en la variable Departamento, existen repetion con la palabra venta.

### 4. Graficas.

### 5. Analizar grupos de atributos

Identificar relaciones entre atributos como dependencia entre ellos. Esta información nos puede llevar más adelante a retirar algunos de los atributos.


### 5.1. Identificar relaciones entre las variables

En este caso se tiene una variable que es el volúmen de producto que se determina a partir de la longitud, anchura y altura, en la *sección 4.5. de este notebook* vimos que existen valores ceros en esta variable lo cual no debería presentarse ya que un producto, siempre debe contar con un volumen,y dados los valores que tienen las otras variables con las que se calcula este valor. Este caso es cuando conocemos las relaciones entre los atributos. Otro análisis, es el de descubrir relaciones entre columnas, que lo revisaremos más adelante en el curso.


In [65]:
# Revisar los datos relacionados
data[data['volumen_cm3_producto']==0][['longitud_cm_producto','altura_cm_producto','ancho_cm_producto','volumen_cm3_producto']]

longitud_cm_producto  altura_cm_producto  ancho_cm_producto  \
6                        6                   5                 16   
7                       51                  32                 20   
14                      17                  16                 45   
15                      30                  23                 19   
29                      10                  46                 29   
...                    ...                 ...                ...   
1535                     0                  35                 36   
1559                    57                  29                 25   
1612                    41                   0                 12   
1622                    34                  13                 31   
1642                    27                   4                 22   

      volumen_cm3_producto  
6                        0  
7                        0  
14                       0  
15                       0  
29                       0  
...                    ...  
1535                     0  
1559                     0  
1612                     0  
1622                     0  
1642                     0  

[126 rows x 4 columns]

En estos casos vemos que es un error de cálculo o digitación ya que para las variables con las que se calcula se tienen valores mayores a cero, por lo cual es necesario modificar el valor de volumen de esos productos.

### 6. Perfilamiento con pandas_profiling

Para terminar, vamos a revisar otra estrategia que nos permite generar informes sobre el perfilamiento de datos, a partir de un DataFrame de pandas. Las funciones que revisamos previamente aunque nos permiten analizar los datos, puede ser básica para un análisis de datos exploratorio más sofisticado. En esos casos, pandas_profiling puede aportar ya que extiende el DataFrame de pandas con df.profile_report () para un análisis rápido de datos.

Para cada columna, genera las siguientes estadísticas, si son relevantes para el tipo de columna, se presentan en un informe HTML interactivo:

1. Inferencia de tipo: detecta los tipos de columnas en un dataframe.
2. Esenciales: tipo, valores únicos, valores faltantes.
3. Estadísticas de cuantiles como valor mínimo, Q1, mediana, Q3, máximo, rango, rango intercuartílico. Esta opción es bastante útil para identificar datos atípicos.
4. Estadísticas descriptivas como media, moda, desviación estándar, suma, desviación absoluta mediana, coeficiente de variación, curtosis, asimetría.
5. Valores más frecuentes.
6. Histogramas.
7. Correlaciones destacando variables altamente correlacionadas, matrices de Spearman, Pearson y Kendall. Esto permite descubrir relaciones entre atributos.
8. Matriz de valores faltantes, recuento, mapa de calor y dendrograma de valores faltantes

Tomado de la librería oficial de pandas_profiling en [github](https://github.com/pandas-profiling/pandas-profiling)

Lo más importante al utilizar esta librería es recordar que lo fundamental son los análisis que hagamos sobre estos reportes.

In [ ]:
## Asegúrese de tener la última versión de pandas profiling instalada
!pip3 install pandas_profiling --upgrade

## Generar objeto de perfilamiento de datos. Esta función puede tardar un par de minutos.
#perfilamiento=pandas_profiling.ProfileReport(data)
#perfilamiento

La invitación es a revisarlo en detalle y mirar cómo lograr los análisis que realizamos en este tutorial.

In [32]:
#Escribir el perfilamiento de los datos en formato html
perfilamiento.to_file("ReportPerfil-ordenes-por-producto.html")

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]